pkg/util/deployment/deployment.go

```diff
@@ -142,6 +131,138 @@ func GetNewReplicaSetFromList(deployment extensions.Deployment, c clientset.Inte
+	// 1. Update all pods managed by the rs to have the new hash label, so they will be correctly adopted.
+	selector, err := unversioned.LabelSelectorAsSelector(rs.Spec.Selector)
+	if err != nil {
+		return nil, err
+	}
+	options := api.ListOptions{LabelSelector: selector}
+	podList, err := getPodList(namespace, options)
+	if err != nil {
+		return nil, err
+	}
+	for _, pod := range podList.Items {
+		// If the pod already has the new hash label, avoid re-labeling it
+		if len(pod.Labels) > 0 && len(pod.Labels[extensions.DefaultDeploymentUniqueLabelKey]) > 0 {
+			continue
+		}
+		pod.Labels = labelsutil.AddLabel(pod.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		delay, maxRetries := 3, 3
+		podName := pod.Name
+		for i := 0; i < maxRetries; i++ {
+			_, err = c.Core().Pods(namespace).Update(&pod)
+			if err == nil {
+				break
+			}
+			time.Sleep(time.Second * time.Duration(delay))
+			delay *= delay
+			getPod, err := c.Core().Pods(namespace).Get(podName)
+			if err != nil {
+				return nil, err
+			}
+			pod = *getPod
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// 2. Update rs label, rs template label, and rs selector to include the new hash label
+	// Copy the old selector, so that we can scrub out any orphaned pods
+	oldSelector := rs.Spec.Selector
+	// Update the selector of the rs so it manages all the pods we updated above
+	updatedRS, err := updateRSWithRetries(c.Extensions().ReplicaSets(namespace), &rs, func(updated *extensions.ReplicaSet) {
+		updated.Labels = labelsutil.AddLabel(updated.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		updated.Spec.Template.Labels = labelsutil.AddLabel(updated.Spec.Template.Labels, extensions.DefaultDeploymentUniqueLabelKey, hash)
+		updated.Spec.Selector = labelsutil.AddLabelToSelector(updated.Spec.Selector, extensions.DefaultDeploymentUniqueLabelKey, hash)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// 3. Clean up any orphaned pods that don't have the new label, this can happen if the rs manager
+	//    doesn't see the update to its pod template and creates a new pod with the old labels after
+	//    we've finished re-adopting existing pods to the rs.
+	selector, err = unversioned.LabelSelectorAsSelector(oldSelector)
+	if err != nil {
+		return nil, err
+	}
+	options = api.ListOptions{LabelSelector: selector}
+	podList, err = getPodList(namespace, options)
+	for _, pod := range podList.Items {
+		if value, found := pod.Labels[extensions.DefaultDeploymentUniqueLabelKey]; !found || value != hash {
+			if err := c.Core().Pods(namespace).Delete(pod.Name, nil); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return updatedRS, nil
+}
+
+type updateFunc func(rs *extensions.ReplicaSet)
+
+func updateRSWithRetries(rsClient unversionedextensions.ReplicaSetInterface, rs *extensions.ReplicaSet, applyUpdate updateFunc) (*extensions.ReplicaSet, error) {
+	var err error
+	oldRs := rs
+	err = wait.Poll(10*time.Millisecond, 1*time.Minute, func() (bool, error) {
```

Comment:
Every client needs to do this. We should move this to the client library. I'll file an issue.